src/crm/activecampaign.ts

```diff
@@ -0,0 +1,300 @@
+          list: listid,
+          contact: contactid,
+          status: 1
+        },
+      }),
+    });
+
+    if (!res.ok) throw new Error(`Failed to subscribe to list: ${res.statusText}`);
+    console.log(`Subscribed contact ${contactid} to list ${listid}`);
+  };
+
+  //The tag must already exist, default?
+  addTagsToContact = async (contactId: string, tagIds: string): Promise<void> => {
+    const ids = tagIds.replace(/\s+/g, "").split(",");
+    for (const tagId of ids) {
+      const res = await fetch(`${url}/api/3/contactTags`, {
+        method: "POST",
+        headers: this.headers,
+        body: JSON.stringify({
+          contactTag: {
+            contact: contactId,
+            tag: tagId,
+          },
+        }),
+      });
+
+    if (!res.ok) {
+      const errorData = await res.json();
+      throw new Error(`Failed to add tag ${tagId}: ${JSON.stringify(errorData)}`);
+    }
+    console.log(`Tag ${tagId} added to contact ${contactId}`);
+  }
+};
+
+  handleActionContact = async (
+    message: ActionMessage
+  ): Promise<handleResult | boolean> => {
+    const email = message.contact.email;
+
+    console.log("Processing action:", message.action.id, "testing:", message.action.testing);
+
+      if (this.verbose) {
+      console.log(JSON.stringify(message, null, 2));
+    }
+    try {
+
+      let camp = await this.fetchCampaign(message.campaign.id);
+      const { listid,
+        tagids,
+        action_id_field,
+        ref_field,
+        data_source,
+        zip_field } = camp.config?.component?.sync || {};
+
+      if (!tagids || !action_id_field || !ref_field) {
+        console.error("Missing required configuration for ActiveCampaign sync");
+        return false;
+      };
+
+      let contactid = await this.fetchContact(email);
+
+      const contactPayload: ContactPayload = {
+        ...(contactid
+          ? pick(message.contact, ['firstName', 'lastName', 'contactRef', 'phone', 'postcode'])
+          : pick(message.contact, ['email', 'firstName', 'lastName', 'contactRef', 'phone', 'postcode'])),
+        id: message.action.id,
+      };
+
+      const bodyContent = this.body(
+        contactPayload,
+        data_source || '29',
+        action_id_field,
+        ref_field,
+        zip_field || '11');
+
+      if (contactid) {
+        console.log("Contact already exists, update:", contactid);
+        contactid = await this.updateContact(contactid, bodyContent);
+      } else {
+        console.log("Creating new contact:", email);
+        contactid = await this.createContact(bodyContent);
+      }
+
+      if (!contactid) {
+        console.error("Failed to create or update contact");
+        return false;
+      }
+
+      // Subscribe to list
+      await this.subscribeToList(contactid, listid || '1');
+
+      // Add petition-specific tags
+      await this.addTagsToContact(contactid, tagids);
+
+      console.log("Action contact processed successfully", message.action.id);
```

Comment:
your crm is quite console verbose, may be some of them should be only called 

   this.verbose && console.log ...

Reply:
I should remove some of the logs, definitely. But I'd still leave a few of them, were useful for debugging, considering it is hard to reproduce whatever edge case might happen,